staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go

```diff
@@ -490,9 +486,46 @@ func (g *Cloud) ensureInternalInstanceGroup(name, zone string, nodes []*v1.Node)
 func (g *Cloud) ensureInternalInstanceGroups(name string, nodes []*v1.Node) ([]string, error) {
 	zonedNodes := splitNodesByZone(nodes)
 	klog.V(2).Infof("ensureInternalInstanceGroups(%v): %d nodes over %d zones in region %v", name, len(nodes), len(zonedNodes), g.region)
+
 	var igLinks []string
-	for zone, nodes := range zonedNodes {
-		igLink, err := g.ensureInternalInstanceGroup(name, zone, nodes)
+	gceZonedNodes := map[string][]string{}
+	for zone, zNodes := range zonedNodes {
+		hosts, err := g.getFoundInstanceByNames(nodeNames(zNodes))
+		if err != nil {
+			return nil, err
+		}
+		names := sets.NewString()
+		for _, h := range hosts {
+			names.Insert(h.Name)
+		}
+		skip := sets.NewString()
+
+		igs, err := g.ListInstanceGroups(zone)
+		if err != nil {
+			return nil, err
+		}
+		for _, ig := range igs {
+			if strings.EqualFold(ig.Name, name) {
+				continue
+			}
+			instances, err := g.ListInstancesInInstanceGroup(ig.Name, zone, allInstances)
```

Comment:
What's the impact of this in terms of cloud provider calls?  `O(Zones * InstanceGroupsPerZone)`, where before it was `O(Zones)`?  Will we hit novel quota / load aspects with this?

Reply:
There is definitely increased API overhead from `O(zones)` -> `O(zones * instanceGroups)`

one way to reduce this would be to use certain labels/name-prefixes to request users to explicitly target discover of the instance-groups by the cluster.

but GCP doesn't support tags for instance-groups yet, https://cloud.google.com/compute/docs/labeling-resources

so our only option would be to restrict to certain prefixes similar to https://github.com/kubernetes/kubernetes/blob/ec344a1c77eea2f9320a340d97dd2de9f49dac3b/staging/src/k8s.io/legacy-cloud-providers/gce/gce_instances.go#L450-L458

Reply:
@smarterclayton i pushed https://github.com/kubernetes/kubernetes/pull/84466/commits/bb6e87f965820ba7e8cc2eb6657fd55f05aad026 to only turn own this external when a prefix is set to make sure we can maintain compatibility in terms of API calls for existing users and keep API calls to minimum when we want to use this new change.